Support using Ollama as an edit prediction provider for FIM models

oliverbarnes opened the pull request:
Closes #15968.

Tested with `Qwen2.5-Coder-3B`, with @boozook's and ghost's (on Zed Discord) help, and `Codellama:7B-code`. Tried `Codestral-22B`, but it ran way too slow on my machine (an Macbook Air M2) to get any meaningful testing done.

Update: later tested with Ollama Cloud, and it worked with Codestral there.

Also tested with Zeta briefly, until realizing it's more sophisticated than FIM. 

<img width="1161" alt="Screenshot 2025-06-29 at 23 06 34" src="https://github.com/user-attachments/assets/878fb4bc-9127-4662-ab06-d426dcb7e7dc" />

<img width="680" height="643" alt="Screenshot 2025-07-17 at 18 17 32" src="https://github.com/user-attachments/assets/36deb8d8-5333-4b73-97cd-a2456e1581fa" />

### Key changes

1. Added Ollama enum variant to EditPredictionProvider
2. Extended Ollama API with `/api/generate` endpoint support, using the [`suffix` param](https://www.postman.com/postman-student-programs/ollama-api/example/21521806-09c37c4b-eca1-419d-ac50-61d466e72e0a) for completion.
3. Implemented OllamaEditPredictionProvider
4. Updated registry to handle Ollama provider assignment and model switching
5. Edit prediction menu supports Ollama API url configuration, model list refresh and model configuration.

The implementation follows (and sometimes merges) patterns used in existing providers (Copilot, Supermaven, Zeta and LM Studio). 

Users can now set `"edit_prediction_provider": "ollama" ` to use local Ollama models for inline code completion*.

Available models are detected and listed in the Edit Predictions menu, and can be locally configured in the settings menu.

Selecting a model in the menu will make Ollama switch to it, and make the configuration for it available in the settings file. 

The menu allows for refreshing the model list, which will update the list shown in the menu while removing any newly unavailable models from the settings file.

*My understanding is that the inline_completion_provider setting was renamed to edit_prediction_provider after issue #15968, which mentions the earlier name, was opened. 

### What this PR does not include

Unfortunately, support for Zeta. I know [this is a desired feature,](https://github.com/zed-industries/zed/discussions/24859#discussioncomment-13610519) but I learned the differences between the different coding models and it became clear Zeta is way more sophisticated and needs special love. Something for a future iteration. (thanks @MonaMayrhofer for clarifying this).

I decided to constrain the scope of the PR to FIM models, as these seem to work out of the box with the `suffix` param for Ollama's /api/generate endpoint (thanks for pointing me to it, @filipwiech).

### Testing it manually

```
git remote add oliverbarnes git@github.com:oliverbarnes/zed.git
git fetch oliverbarnes
git checkout -b ollama-inline-completions
cargo build --release
./target/release/zed
```

```json
"features": {
    "edit_prediction_provider": "ollama"
},
"language_models": {
    "ollama": {
        "api_url": "http://localhost:11434",
        "available_models": [
        {
            "name": "qwen2.5-coder:3b",
            "display_name": "Qwen 2.5 Coder 3b",
            "max_tokens": 4096,
            "keep_alive": null,
            "supports_tools": null,
            "supports_images": null,
            "supports_thinking": null
        },
        {
            "name": "codellama:7b-code",
            "display_name": "Codellama 7B code",
            "max_tokens": 4096,
            "keep_alive": null,
            "supports_tools": null,
            "supports_images": null,
            "supports_thinking": null
        },
        ]
    }
}
```

### Potential future improvements
- Zeta support

### TODO
- [x] Edit prediction menu
- [x] Ollama api key support
- [x] Restrict api url config to ollama
- [x] Eager / subtle support
- [x] Partial accepts
- [x] Don't update prediction counter
- [x] Test with coding completion models other than Qwen 2.5 coder 3B (confirmed working with other FIM models)
- [x] ~~Dedicated handling for Zeta~~ (sticking to FIM models in this PR)
- [x] ~~Log completion errors~~ (decided to follow other completion providers, which don't log errors. Copilot sends telemetry upstream)
- [x] Documentation
- [x] Improve tests (not testing the menu itself, as other providers don't either - might try adding tests for it after this PR)
- [x] codellama::7b now 400ing for some reason (looks like only codellama:7b-code supports FIM)
- [x] Log when we requests can't be sent to Ollama for some reason, as when we forget to run `ollama serve` beforehand
- [x] Automatic detection of available Ollama models

Release Notes:

- Added support for Ollama as an edit prediction provider, constrained to FIM models for now.

oliverbarnes committed:
Support using ollama as an inline_completion_provider

oliverbarnes committed:
Remove unused dependency on multibuffer

oliverbarnes commented:
Got completion going with Ollama + codellama 🎉 Added a screenshot to the PR description.

Man, it's a rush seeing the locally built zed running for the first time with a feature you added yourself 🏄‍♂️ 

Now on to testing with deepseek-coder, zeta and others. One kink I found is that, on the ollama completions menu at the bottom right, clicking on Ollama Settings just opens http://localhost:11434/ on the browser.

Toggle Ollama Completions seems to toggle off (no requests are sent to the Ollama server), but I'm not sure whether that's the way it's supposed to work or whether it should toggle with another provider - need to readup on this.

```
{
  "show_edit_predictions": true,
  "features": {
    "edit_prediction_provider": "ollama"
  },
  "edit_predictions": {
    "mode": "subtle",
    "provider": "ollama"
  },
  "language_models": {
    "ollama": {
      "api_url": "http://localhost:11434",
      "available_models": [
        {
          "name": "codellama:latest",
          "display_name": "CodeLlama Latest",
          "max_tokens": 4096
        },
        {
          "name": "deepseek-coder:latest",
          "display_name": "DeepSeek Coder Latest",
          "max_tokens": 4096
        }
      ]
    }
  }
```

oliverbarnes commented:
Tested with [Zeta as well](https://github.com/zed-industries/zed/discussions/24859#discussioncomment-13612949)

oliverbarnes commented:
@Razboy20 thanks! But so far as I understand it, inline assistant (the inline chat opened with ctrl-enter) and edit prediction (suggested inline completions) are separate features that can be configured differently? 

I'm basing myself on https://zed.dev/docs/ai/edit-prediction#github-copilot and https://zed.dev/docs/ai/configuration#feature-specific-models

So far I haven't come across any docs or code suggesting edit prediction having to defer to the inline assistant model choice if different. But I'm new to the codebase, so if there is, please let me know!

Razboy20 commented:
@oliverbarnes oh whoops, you're totally right. Either way, might be good to add some configurability and error reporting if inline completions fail.

oliverbarnes commented:
Definitely, I'm currently wrapping my mind around what options the edit prediction can show, and whether it varies for different providers.

How would is error reporting handled with other providers? Haven't looked into it yet

oliverbarnes committed:
First stab at the prediction menu

oliverbarnes committed:
Prediction menu more fleshed out but not quite there yet

Connection status works on second open of menu, api url configurable, model switcher not working yet

oliverbarnes committed:
Actually, a lot of the options aren't needed

Remove connection status, refresh connection and the ollama web ui options

oliverbarnes committed:
Ollama model switcher working

oliverbarnes commented:
Model switching now working on the prediction menu:

<img width="595" alt="Screenshot 2025-07-02 at 13 32 58" src="https://github.com/user-attachments/assets/4958cf87-9632-4248-8a3b-95973898c480" />

boozook commented:
That is so cool! Now I see Zed as potentially day-to-day instrument for me. I hope this PR will be merged soon! 😍

oliverbarnes committed:
Improved FIM token handling per model

oliverbarnes committed:
Appease clippy

filipwiech commented on crates/ollama/src/ollama_completion_provider.rs:
```diff
@@ -0,0 +1,786 @@
+
+        if model_lower.contains("qwen") && model_lower.contains("coder") {
+            // QwenCoder models use pipes
+            format!("<|fim_prefix|>{prefix}<|fim_suffix|>{suffix}<|fim_middle|>")
```
I could be wrong, but I'm pretty sure that it's Ollama's job to abstract over this kind of internal model details. Basically each model's template should already define and use the correct tokens. Looks like Ollama provides support for the FIM (fill-in-the-middle) completion in their API via the optional `suffix` parameter, see [the documentation](https://github.com/ollama/ollama/blob/main/docs/api.md#generate-a-completion). :slightly_smiling_face:

oliverbarnes replied:
Thanks! I was hoping for predictions to just work initially, and thought I had it working at one point, but then some FIM tokens started showing up in predictions. Hence I started going down this rabbit hole of handling the different cases... And just learned that it's a whole different ballgame with Zeta.

So it'd be a relief not to worry about this - gonna read up on those docs.

oliverbarnes replied:
Ok, removed the FIM token handling and introduced the suffix param. But now I'm getting some chat completion and not just code :P 

need to spend more time grokking this.

oliverbarnes committed:
Use Ollama's suffix field and remove FIM token handling

oliverbarnes commented:
> I might be misunderstanding something here, but as this PR mentions using zeta as an edit prediction provider, I think it should be noted, that - as far as I can tell - the zeta model is not fine tuned to using the standard FITM approach

No, you're absolutely right. I found out the hard way just yesterday as I was fiddling with the FIM tokens.

I started testing this with Zeta after seeing demand for it on another discussion, and naïvely assumed it'd use FIM like other coding models, specially Qwen coder.

I'm now just getting familiar with it, and your explanation and example help with that. I'm midway watching the video where it's explained how Zeta works. It's real interesting

> Of course this is specific to just the zeta model (and maybe similar alternatives to come in the future?) and not all ollama edit prediction providers, but if we want to leverage it's full power in a self-hosted environment it should probably be handled accordingly.

Agreed. I'm realizing it's too big a bite for this PR :) I think I'll restrict the scope here to FIM models, and continue looking into Zeta afterwards. 

I'm beginning to suspect it might require the knowledge of the Zed team, though.

mspinelli commented:
Also, something that would be good to double-check is to make sure this doesn't increment the online prediction completion counter when running locally. 

I mention this because an alternative way to run locally, for example, https://github.com/g0t4/zed-zeta-server, proxies locally using ZED_PREDICT_EDITS and this nonetheless will increment the counter when edits are accepted because I suspect zed's code manages that.

mspinelli commented:
> I might be misunderstanding something here, but as this PR mentions using zeta as an edit prediction provider, I think it should be noted, that - as far as I can tell - the zeta model is not fine tuned to using the standard FITM approach (because it doesn't only fill in at the location of the cursor, but instead, it can predict edits all across the file).
> 
> Judging by their [dpo finetuning notebook](https://huggingface.co/datasets/zed-industries/zeta/blob/main/script/dpo.ipynb) it looks like they are instead using a prompt like this:
> zeta model prompt
> 
> And then the model responds like this
> zeta model output
> 
> Of course this is specific to just the zeta model (and maybe similar alternatives to come in the future?) and not all ollama edit prediction providers, but if we want to leverage it's full power in a self-hosted environment it should probably be handled accordingly.

Tangential to this discussion, but given those finetuning notebooks, I wonder if it would be possible to create a 1.5B or 3B model version of Zeta to speed up local inference on Macs. Or would those models be too small to be useful for the kind of predictions Zeta makes?

oliverbarnes commented:
> Also, something that would be good to double-check is to make sure this doesn't increment the online prediction completion counter when running locally.
> 
> I mention this because an alternative way to run locally, for example, https://github.com/g0t4/zed-zeta-server, proxies locally using ZED_PREDICT_EDITS and this nonetheless will increment the counter when edits are accepted because I suspect zed's code manages that.

Thanks @mspinelli, I wasn't aware of the counter. Gonna look into it

oliverbarnes committed:
Remove stop tokens and completion cleanup

oliverbarnes committed:
Support using an API key

oliverbarnes commented:
API key support added - but needs to be tested with a remote Ollama server

oliverbarnes committed:
Ensure only Ollama's api url is configurable

oliverbarnes committed:
Eager / subtle now working

oliverbarnes committed:
Enable partial acceptance

oliverbarnes committed:
Don't update edit prediction counter during Ollama predictions

And during predictions by other third-parties.

No test coverage - haven't found a way to test that the telemetry event isn't sent, yet

oliverbarnes commented:
> Also, something that would be good to double-check is to make sure this doesn't increment the online prediction completion counter when running locally.

@mspinelli just submitted a fix for this, and it seems to be working for me. When you get a chance would you please confirm on your end?

No test coverage, though. Only way I found was by initializing the telemetry queue, creating a separate channel... Felt too heavy handed. I might be missing something, but I didn't find telemetry events tested elsewhere in the codebase

oliverbarnes committed:
Documentation

oliverbarnes committed:
Default to Qwen Coder

oliverbarnes committed:
Document the API url setting

boozook commented:
Imho, there shouldn't be default value, but required one for model in the https://github.com/zed-industries/zed/commit/fa5e7c4631976957cfccdfe81ef4ea6c9b8c55e8

oliverbarnes commented:
> Imho, there shouldn't be default value, but required one for model in the [fa5e7c4](https://github.com/zed-industries/zed/commit/fa5e7c4631976957cfccdfe81ef4ea6c9b8c55e8)

Thanks, I am a bit on the fence regarding having a default value or not. I figured it could be good for someone not familiar with FIM models to get up and running, but yeah, a little too prescriptive perhaps.

Not sure I understand the second part of what you said, though? Required one for model...?

boozook commented:
I only meant that instead of the default value, there could be a required value.

oliverbarnes committed:
Remove model default and add placeholder configuration link

oliverbarnes committed:
Remove current inline completion button tests, they aren't very useful

oliverbarnes committed:
Let's start over with tests

oliverbarnes committed:
Ollama::fake

oliverbarnes committed:
Test completion flow

oliverbarnes committed:
Test partial typing

oliverbarnes committed:
Updated dev dependencies

oliverbarnes committed:
Test partial acceptance of completion suggestion

oliverbarnes committed:
Test completion invalidation

oliverbarnes commented:
Now testing with codellama:7b-code, after learning that `7b` doesn't really support FIM. 

Suggestions appear consistently, but at times the stop token <EOT> shows up. I'm surprised this is happening with a Meta model, I'd have expected it to work out of the box with Ollama. Meaning, I'd have expected Ollama to strip these away for Codellama

oliverbarnes committed:
Send codellama:7b-code stop token in request

So Ollama filters it out

oliverbarnes commented:
Now sending a `stop` options param to Ollama with the stop token so it knows not to send it. I was hoping not to deal with any template tokens in the implementation, but Ollama really doesn't seem to handle this behind the scenes, from what I could tell.

oliverbarnes committed:
Updated top level Cargo.lock

oliverbarnes commented:
Running test suites for the affected crates all pass except for `editor`, where 55 tests fail. Ran some of them individually and they pass. Also checked running `editor`'s test suite on `main`, seems like the same problem happens there. I'm guessing it's a known build problem?

Otherwise, I think this PR is ready for a review :) 🙏. Opening it up

```
failures:
    editor_tests::test_on_type_formatting_is_applied_after_autoindent
    git::blame::tests::test_blame_random
    hover_popover::tests::test_empty_hovers_filtered
    hover_popover::tests::test_hover_inlay_label_parts
    hover_popover::tests::test_info_hover_with_hrs
    hover_popover::tests::test_keyboard_hover_info_popover
    hover_popover::tests::test_line_ends_trimmed
    hover_popover::tests::test_mouse_hover_info_popover
    hover_popover::tests::test_mouse_hover_info_popover_with_autocomplete_popover
    inlay_hint_cache::tests::test_basic_cache_update_with_duplicate_hints
    inlay_hint_cache::tests::test_cache_update_on_lsp_completion_tasks
    inlay_hint_cache::tests::test_excerpts_removed
    inlay_hint_cache::tests::test_hint_request_cancellation
    inlay_hint_cache::tests::test_hint_setting_changes
    inlay_hint_cache::tests::test_inlays_at_the_same_place
    inlay_hint_cache::tests::test_inside_char_boundary_range_hints
    inlay_hint_cache::tests::test_large_buffer_inlay_requests_split
    inlay_hint_cache::tests::test_multiple_excerpts_large_multibuffer
    inlay_hint_cache::tests::test_no_hint_updates_for_unrelated_language_files
    inlay_hint_cache::tests::test_toggle_inlay_hints
    inline_completion_tests::test_inline_completion_insert
    inline_completion_tests::test_inline_completion_invalidation_range
    inline_completion_tests::test_inline_completion_jump_button
    inline_completion_tests::test_inline_completion_modification
    inline_completion_tests::test_partial_accept_inline_completion
    items::tests::test_deserialize
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_autocloses_dot_separated_component
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_autocloses_fragment
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_autocloses_in_jsx_expression
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_basic
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_basic_ignore_already_closed
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_basic_nested
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_does_not_autoclose_comment
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_does_not_autoclose_doctype
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_does_not_include_type_argument_in_autoclose_tag_name
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_doesnt_autoclose_already_closed_in_jsx_expression
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_doesnt_autoclose_closing_tag
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_doesnt_autoclose_on_gt_in_expr
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_doesnt_mess_up_trailing_text
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_ignores_closing_tags_in_expr_block
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_ignores_closing_tags_with_different_tag_names
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_jsx_attr
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_multi_cursor_autoclose_different_tags
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_multi_cursor_autoclose_same_tag
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_multi_cursor_autoclose_some_dont_autoclose_others
    jsx_tag_auto_close::jsx_tag_autoclose_tests::test_multibuffer
    mouse_context_menu::tests::test_mouse_context_menu
    movement::tests::test_find_boundary
    movement::tests::test_find_preceding_boundary
    movement::tests::test_find_preceding_boundary_with_inlays
    movement::tests::test_move_up_and_down_with_excerpts
    movement::tests::test_next_subword_end
    movement::tests::test_next_word_end
    movement::tests::test_previous_subword_start
    movement::tests::test_previous_word_start
```

oliverbarnes committed:
Cleanup a couple of tests

oliverbarnes committed:
Log a warning when ollama isn't available

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

# Conflicts:
#	Cargo.lock

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes commented:
While the PR is the queue, I'm going ahead and working on automatic detection of available models.

oliverbarnes committed:
Auto detect models WIP

oliverbarnes committed:
Selected model now saved to settings file

oliverbarnes committed:
Cleanup unavailable models from settings

oliverbarnes commented:
Automatic detection of models done. Testers welcome :)

- Available models are detected and listed in the Edit Predictions menu, and can be locally configured in the settings menu.
- Selecting a model in the menu will make Ollama switch to it, and make the configuration for it available in the settings file. 
- The menu allows for refreshing the model list, which will update the list shown in the menu while removing any newly unavailable models from the settings file.

maxdeviant commented on crates/editor/src/editor.rs:
```diff
@@ -7414,6 +7414,12 @@ impl Editor {
 
+        // Only report telemetry events for online providers managed by Zed's service,
+        // not for third-party providers like Ollama that don't count toward online usage
+        if provider.name() != "zed-predict" {
```
I don't think we want to do this. We collect telemetry for all providers for informational purposes (so we can see which ones are most used).

The usage counting/limiting for Zed's Edit Predictions is handled separately from telemetry.

boozook replied:
@maxdeviant,
I suppose that what you offer will lead to the fact that more people will refuse telemetry, at least. The next logical step is to add a list of everything found in `$PATH` to telemetry?

oliverbarnes replied:
Thanks, makes sense - I wasn't aware the counter was a separate thing. Reverted

oliverbarnes committed:
Revert "Don't update edit prediction counter during Ollama predictions"

This reverts commit ed412b5871cbf944df9ba40176b1a87f8c3f1c7a.

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

# Conflicts:
#	docs/src/ai/configuration.md

oliverbarnes committed:
Update docs with automatic detection of models and model switching

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

# Conflicts:
#	Cargo.lock
#	crates/zed/src/zed/inline_completion_registry.rs

oliverbarnes committed:
Add new user_agent method required by updated HttpClient trait

oliverbarnes commented:
Thanks, makes sense. And for sure, not going to remove the settings :) 

I'll look into merging custom settings in then, in place of the current automatic cleanup.

@maxdeviant please chime in if the team has a different preference

oliverbarnes committed:
Revert "Cleanup unavailable models from settings"

This reverts commit 0f135387523a604930b2c910bd7c0023e19093c9.

oliverbarnes committed:
Merge models in local settings with ones listed by ollama

This allows for the scenario where the user doesn't have access to ollama's listing and needs to tell Zed explicitly, by hand

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

# Conflicts:
#	crates/zed/src/zed/inline_completion_registry.rs

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

# Conflicts:
#	Cargo.lock
#	crates/edit_prediction_button/src/edit_prediction_button.rs
#	crates/editor/src/editor.rs
#	crates/zed/src/zed/edit_prediction_registry.rs

oliverbarnes commented:
Working on conflicts - gonna refactor this PR's legacy inline_completion names to edit_prediction ones accordingly.

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

# Conflicts:
#	crates/edit_prediction_button/src/edit_prediction_button.rs
#	crates/editor/src/editor.rs

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

# Conflicts:
#	crates/zed/src/zed/edit_prediction_registry.rs

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

ghost91- commented:
Hey @oliverbarnes,

I've just recently implemented the `/api/tags` API in my company's custom LLM server and wanted to try out if automatic model detection works now.

Unfortunately, there is still a problem: Zed doesn't send the `OLLAMA_API_KEY` with the request to `/api/tags`. So our LLM server rejects that request with a 401, so autodetection does not work.

Could you adapt it so that the `OLLAMA_API_KEY` is also sent as BVearer token with that request (and any other requests to the Ollama server, that might also be made, not sure if you use the `/api/show` endpoint, for example, to detect model capabilities...)?

ohroy commented:
hope to merge. zed only have 2000 times now, cant use windsurf cant use ollama like qwen-coder, the only way to contiune is ....

oliverbarnes committed:
Send api key for api/show and api/tags

oliverbarnes committed:
Log error instead of warning when ollama server is unvailable

oliverbarnes committed:
Remove unneeded comments

oliverbarnes committed:
Remove default ollama prediction model setting

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes commented:
@ghost91- I've pushed a fix. I'm in the process of setting up a VPS with an Ollama instance so I can test it on my own, but am a little short of time until middle of next week. Meanwhile, if you get a chance to test yourself, would you let me know how it goes? Else I'll post back the results here.

Also submitted a couple of other fixes / improvements. No more qwen coder default, and we're now logging an error instead of a warning when ollama isn't available.

@maxdeviant a note about the api key support - to avoid (further) scope creep I've only implemented it for Ollama edit predictions, not for the agent or inline assistant.

ghost91- commented:
@oliverbarnes I'm not exactly sure why, but it still doesn't work (reliably) for me. I have seen some requests to `/api/tags`, in which the api key is included. But I cannot reproduce it reliably. In most requests that I see, it's not included. I have not been able to see a model list yet.

ghost91- commented:
By the way, there is also a separate PR that aims to add support for API keys for ollama: https://github.com/zed-industries/zed/pull/34110. Maybe it makes sense to join forces, to avoid duplicating the effort?

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

# Conflicts:
#	crates/zed/src/zed/edit_prediction_registry.rs

oliverbarnes committed:
Rename OllamaService to State

To follow the convention used in language_models

oliverbarnes committed:
Use OLLAMA_API_KEY_VAR to store env var string

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes commented:
I spent some time looking into how to best integrate with the api key support added in #34110. 

Tried sharing authentication between ollama edit prediction and language model (agent/inline assistant), but quickly got tied into coupling and circular dependencies. 

So I had a closer at the different provider crates and `language_models` provider files, and I think I have a better sense of the architecture. My current mental model is that the provider crates have "core" provider logic, and `language_models` are adapters for them into Zed, implementing a common provider interface for agent and inline assistant implementations.

Edit predictions is its own thing, with their Zed interface defined the `edit_prediction` crate, but provider implementations in the core `copilot`, `supermaven` and, with #15968, `ollama` crates. I have the impression they could be moved to an `edit_predictions` crate, mirroring `language_models` (which in my mind now is more like "assistants"), so there'd be a more clear separation between "core" and Zed AI integration adapter crates.

In the end I realized Ollama edit predictions and agent/inline assistant don't really need to share authentication like copilot's do. Copilot is more complicated, with oauth and both an LSP and an API to interact with, while Ollama just needs to inject an api key in requests. Edit prediction and assistant providers can each use the api key separately.

It'd just be good to avoid duplication of the api key env var constant, so I'm proposing it live in core `ollama` instead of language_models (as it is now in #34110). 

I've also renamed the edit prediction provider's `OllamaService` to `State`, to mirror the naming in `language_models`.

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes commented:
@ThePerfectComputer thanks! I like both the idea of enabling ollama predictions through the command palette and of showing request errors in the status bar. I'd be happy to work on those too - I'm curious how the command palette works, for example.

> Inline editing doesn't work with non FIM models at the moment.

Indeed the scope of the PR has been restricted to FIM models after starting out with a naïve attempt to make it work for other types of models too, including Zeta.

I did play with filtering out non-FIM available models, but didn't find another way besides testing each model to see if the request errored or not, when submitting a suffix. In the end some community feedback here in the PR led me to assume that users deciding to go with Ollama would know what models they'd like to use specifically, and that FIM was required for completion, and that it was better not to do any pre-filtering on available models. That said, I'd also be open to re-discuss this. Let me know if you know of a better way to detect FIM models, for instance? Keeping a hard-coded list also doesn't seem great to me.

**Here's the overriding issue for me, though**: scope creep. This PR is already sizable and long running, and so far it hasn't been validated by the Zed team. I'm just a contributor. 

Even the desire to implement Ollama predictions is an assumption, based on the `Feature` issue type added by a team member to the [original feature request](https://github.com/zed-industries/zed/issues/15968), and nothing else besides apparent community demand.

I'd be more comfortable further discussing and possibly adding these features (or merging commits from you if you're itching to tackle them) after some feedback from them first. After a few nudges, I'm wary of insisting - I suspect they're overwhelmed with their internal backlog.

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

syncdoth commented:
Loving this feature!

One suggestion: could we add some debouncing? In zed, every keystroke invokes the `/predict_edits` request, and this is overloading the ollama server (I'm testing with M4 Pro). 

e.g. Suppose I was typing `"print("`. This sends 6 requests to ollama server, gets queued up, and the final request -- which is basically the only request I care about -- end up taking 16 seconds.

```
[GIN] 2025/08/27 - 23:56:26 | 200 |  3.706427542s |       127.0.0.1 | POST     "/api/generate"
[GIN] 2025/08/27 - 23:56:28 | 200 |  4.371998041s |       127.0.0.1 | POST     "/api/generate"
[GIN] 2025/08/27 - 23:56:31 | 200 |  6.969691041s |       127.0.0.1 | POST     "/api/generate"
[GIN] 2025/08/27 - 23:56:33 | 200 |  9.356582125s |       127.0.0.1 | POST     "/api/generate"
[GIN] 2025/08/27 - 23:56:38 | 200 | 13.901358417s |       127.0.0.1 | POST     "/api/generate"
[GIN] 2025/08/27 - 23:56:41 | 200 | 16.035952208s |       127.0.0.1 | POST     "/api/generate"
```

oliverbarnes commented:
> One suggestion: could we add some debouncing? In zed, every keystroke invokes the `/predict_edits` request, and this is overloading the ollama server (I'm testing with M4 Pro, codellama:code).

Hmm, debouncing _is_ implemented, or at least I thought it was 🤦‍♂️ but I'm seeing the same issue now. 

I think you may have found a bug. Thanks for letting me know @syncdoth! Working on a fix

ooJaan commented:
Since we only care about the last request, how about aborting all the old requests to avoid unnecessary compute and latency?

oliverbarnes committed:
Add test for debouncing

oliverbarnes committed:
Move generate option values into constants

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Test slow typing

oliverbarnes commented:
@syncdoth looking closer into it, I think debouncing is working as expected and we're just not typing fast enough - debouncing is set to 75ms (same as Copilot's). 

I've added a couple of tests asserting this - one tests that only one request is sent after three refreshes done in under 75ms, and another tests slower sequences of refreshes, when a request does get sent for each.

Perhaps this is something that could be configurable down the line, like LSP debouncing (though I wouldn't add it in this PR, and should be consistent across providers). I will continue testing with longer debounces to see how it goes as well, and will report back here.

ThePerfectComputer commented:
> I'd be more comfortable further discussing and possibly adding these features (or merging commits from you if you're itching to tackle them) after some feedback from them first. After a few nudges, I'm wary of insisting - I suspect they're overwhelmed with their internal backlog.

@oliverbarnes I think you're right. It could be the most important thing to get this merge without feature creep. I may still toy around with adding more features locally.

Zed seems aware of which models support tool calling for chat completion. I'm not sure if this is hardcoded or reported in Ollama's `api/tags` endpoint. Will require further investigation.

For FITM models, I'm not seeing `api/tags` report FITM support.

I'm also curious about other models that support the diff format Zeta offers... Will look into that as well.

syncdoth commented:
@oliverbarnes thanks for a quick follow up!

Compared to other local edit prediction provider (e.g. Continue/Twinny for vscode), current implemetation seems to overload the server a lot more and slow it down. Probably they have some server-side logic on top of debouncing, I gotta dig deeper into that.

techy-0 commented:
I know I am echoing an opinion of somebody who came earlier, but wouldn't
using a combination of debouncing, which we can leave at 75ms, and
canceling outstanding completion requests fix all of this issues? I may be
overlooking something but this seems like a simple, yet elegant solution
without adventuring into scope-creep.

On Thu, Aug 28, 2025 at 6:28 PM Sehyun Choi ***@***.***>
wrote:

> *syncdoth* left a comment (zed-industries/zed#33616)
> <https://github.com/zed-industries/zed/pull/33616#issuecomment-3235133467>
>
> @oliverbarnes <https://github.com/oliverbarnes> thanks for a quick follow
> up!
>
> Compared to other local edit prediction provider (e.g. Continue/Twinny for
> vscode), current implemetation seems to overload the server a lot more and
> slow it down. Probably they have some server-side logic on top of
> debouncing, I gotta dig deeper into that.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/zed-industries/zed/pull/33616#issuecomment-3235133467>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACD5AYWMZ5WI5G2R5VVESDD3P57ADAVCNFSM6AAAAACAML63GCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTEMZVGEZTGNBWG4>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

techy-0 commented:
@oliverbarnes - I have only used one debouncing library and it was for web front-end, so this may not be universal; it seemed like it would wait for the events (key presses in both cases) to stop for the debounce time before it would call the function, but once the function is called, it is out of the debouncing library's hands. We had an API endpoint that would take 1-2 seconds to get a response, but I could see it sending more requests whiling leaving the request active. This really started to rear its head when the second request only took 1 second, while the first request may have taken 2 seconds, resulting in the response to the first request getting rendered instead of the last request. In the end we had to keep track of the outstanding request and cancel it before sending a new request.

Again, this is based on one web front-end library, so your mileage may very.

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

seivan commented:
Instead of setting the provider can’t you supply the API address instead so it could work with other applications? Like say LM Studio?

oliverbarnes commented:
A generic openai-compatible provider is an interesting idea, I've seen it implemented elsewhere (see Tensorzero), but I don't know how the suffix part works there - FIM predictions depend on it

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes commented:
> Messages
> 📖	
> This PR includes links to the following GitHub Issues: #[continuedev/continue#3852](https://github.com/continuedev/continue/issues/3852) If this PR aims to close an issue, please include a `Closes #ISSUE` line at the top of the PR body.
> 
> Generated by 🚫 [dangerJS](https://danger.systems/js) against [f04791e](https://github.com/zed-industries/zed/commit/f04791e84846f758566d47bb6831246002d0c82a)

Done.

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

mgsloan committed:
Consolidate ollama settings structs + fix clippy

mgsloan committed:
Remove context field from generate request / response

It's not used and makes dbg! output large and [is deprecated](https://github.com/ollama/ollama/issues/10576)

mgsloan committed:
Remove unused `text` dependency in ollama/Cargo.toml

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Remove regex settings check

oliverbarnes committed:
Remove API url configuration option altogether

oliverbarnes committed:
Auto-refresh on menu open

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Consolidate Ollama model discovery

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Extract ollama_edit_predictions crate

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

Moxinilian commented on crates/ollama_edit_predictions/src/ollama_edit_prediction_provider.rs:
```diff
@@ -0,0 +1,1340 @@
+        // if the text before the cursor ends with a prefix of our completion
+        let mut prefix_len = 0;
+        for i in 1..=completion_text.len().min(text_before_cursor.len()) {
+            if text_before_cursor.ends_with(&completion_text[..i]) {
```
Slices on strings operate at the byte level, not the character level. If there is any non-ascii unicode in the completion, this will crash, as the slicing will eventually set the boundary within a code point.

mgsloan replied:
I think this can be done without materializing the buffer text before the cursor as a string or search it.  If the text before the cursor gets changed (say, by a collaborator), then it could cause a much shorter prefix to match and then cause a much larger, shifted prediction.

Instead it should store anchors for the completion range and do something similar to the [copilot implementation](https://github.com/zed-industries/zed/blob/1c09985fb3a8ab43c251f53c0968cb01e7d41335/crates/copilot/src/copilot_completion_provider.rs#L232)

mgsloan commented:
Sorry, haven't gotten around to properly returning to review of this.  However, some updates:

* #34110 was merged
* Added new excerpt selection logic [here](https://github.com/zed-industries/zed/blob/main/crates/edit_prediction_context/src/excerpt.rs).  It is not yet used in Zeta, but it is planned to soon.  Feel free to use it here - it should be called with `include_parent_signatures: false`.

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

# Conflicts:
#	crates/language_models/src/provider/ollama.rs
#	crates/ollama/src/ollama.rs

oliverbarnes commented:
@mgsloan No worries, I'm working on merging in the api key changes and will integrate the new excerpt selection after that.

One thing I noticed was that the API key constants were moved back into `language_models`? We had moved it to `ollama` to be shared across the former and `ollama_edit_predictions` crate, while avoiding circular dependencies between `language_models` and `ollama` crates - as in `ollama` depending on `language_models` to get the api key, and `language_models` depending on `ollama` for the api interaction helpers that use the api key ♾️. I ellaborate on this [comment](https://github.com/zed-industries/zed/pull/33616#issuecomment-3218167956). 

I'm assuming moving them back was deliberate, but then I might not be understanding the code layout well :)

oliverbarnes committed:
Adapting to ApiKeyState changes

oliverbarnes committed:
Re-use language_models ollama provider facilities

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes commented:
Quick update:
- Dried up the ollama edit prediction provider, keeping only prediction-specific code and otherwise delegating to the ollama provider code in `language_models`.
- Merged in the new api key management, there were some conflicts, now testing it out.
- Reproduced the bug @Moxinilian pointed out. Gonna implement the anchor-based solution next.

oliverbarnes committed:
Improve handling of settings change

oliverbarnes committed:
Use anchors to avoid materializing text

oliverbarnes committed:
Incorporate EditPredictionExcerpt

oliverbarnes committed:
Remove debug logging

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

# Conflicts:
#	Cargo.lock
#	crates/language/src/language_settings.rs
#	crates/language_models/src/provider/ollama.rs
#	crates/ollama/src/ollama.rs

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Fix warning

oliverbarnes committed:
Fix test

oliverbarnes committed:
Fix more tests

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

# Conflicts:
#	crates/language_models/src/provider/ollama.rs

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Missed rename of supports_vision -> supports_images

oliverbarnes committed:
Fix doc formatting

oliverbarnes committed:
Move test imports down into the tests mod

oliverbarnes committed:
Forgot to run clippy, apologies

mgsloan commented on crates/editor/src/edit_prediction_tests.rs:
```diff
@@ -562,3 +562,36 @@ impl EditPredictionProvider for FakeNonZedEditPredictionProvider {
     }
 }
+
+#[gpui::test]
```
This should go next to the other tests

mgsloan commented on crates/edit_prediction_button/src/edit_prediction_button.rs:
```diff
@@ -1009,3 +1258,730 @@ fn toggle_edit_prediction_mode(fs: Arc<dyn Fs>, mode: EditPredictionsMode, cx: &
 }
+
+#[cfg(test)]
+mod tests {
```
Leaning towards deleting these tests.  There is just not that much code under tests and it would be very obvious in the UI if it was broken.  Zed doesn't typically have fine grained UI unit tests.

I didn't look all that closely, but it seems like `test_ollama_menu_shows_discovered_models` and `test_ollama_menu_shows_service_models` are testing the same thing.  Those tests also don't appear to be testing code in this module.  The other tests all seem pretty similar and it isn't really testing functionality of this module other than the button not crashing.  Could have one test that does a few things - but it should test outputs of the code in this module.

mgsloan commented on crates/ollama_edit_predictions/src/ollama_edit_prediction_provider.rs:
```diff
@@ -0,0 +1,1134 @@
+    }
+
+    #[gpui::test]
+    async fn test_completion_invalidation(cx: &mut TestAppContext) {
```
Delete - this is testing details of the editor implementation.  Yeah, copilot test does that but it probably shouldn't

mgsloan commented on crates/ollama_edit_predictions/src/ollama_edit_prediction_provider.rs:
```diff
@@ -0,0 +1,1134 @@
+
+            (prefix, suffix, Some(excerpt.range))
+        } else {
+            // Fallback to original simple extraction
```
Delete

mgsloan commented on crates/zed/src/zed/edit_prediction_registry.rs:
```diff
@@ -3,14 +3,49 @@ use collections::HashMap;
+            log::info!("Using local Ollama instance at: {}", api_url);
+        }
+
+        let task = provider.update(cx, |provider, cx| provider.authenticate(cx));
```
Ollama authenticate should not be called on zed init, only if it's actually be configured to be used.

mgsloan commented on crates/zed/src/zed/edit_prediction_registry.rs:
```diff
@@ -242,5 +291,278 @@ fn assign_edit_prediction_provider(
+                Some(first_model.name.clone())
+            } else if let Some(provider) = OllamaLanguageModelProvider::global(cx) {
+                // Use first discovered model
+                provider
```
This should probably have some heuristics for identifying known FIM models.

oliverbarnes replied:
When looking into originally I didn't find a way other than actually sending requests to ollama to check on models. But I'll take another look

oliverbarnes replied:
@mgsloan the list of models [returned by the Ollama API](https://docs.ollama.com/api#list-local-models) doesn't yet include any information on whether models support FIM or not, so besides this request we'd need to send individual [`/api/show` requests](https://docs.ollama.com/api#show-model-information) for each of the listed models to get their template info.

Is this acceptable? Right now Ollama Cloud (previously Turbo) doesn't return a huge list of models, but in the future I have no idea how many they'll support - my guess would be many.

Actually had this implemented a couple of months ago, but thought it better to leave it to the user to define the models they wanted. The agent seems list all models too, though not all models are great for agents.

Let me know and I'll re-implement, I don't really have a strong opinion. Just wanted to double check before doing so. 

I continue with the clean up and build fixing.

mgsloan commented on crates/ollama_edit_predictions/src/ollama_edit_prediction_provider.rs:
```diff
@@ -0,0 +1,1134 @@
+        let fake_http_client = Arc::new(ollama::fake::FakeHttpClient::new());
+
+        // Mock /api/tags response (list models)
+        let models_response = serde_json::json!({
```
Simplify the response

mgsloan commented on crates/ollama_edit_predictions/src/ollama_edit_prediction_provider.rs:
```diff
@@ -0,0 +1,1134 @@
+                OllamaLanguageModelProvider::api_url(cx).to_string(),
+            )
+        } else {
+            // Fallback if global service isn't available
```
Delete

mgsloan commented on crates/ollama/src/ollama.rs:
```diff
@@ -357,6 +367,108 @@ pub async fn show_model(
 }
 
+#[cfg(any(test, feature = "test-support"))]
+pub mod fake {
```
I'd prefer use of the more general FakeHttpClient.  I am not sure why I don't see more use of it.

oliverbarnes committed:
Remove edit prediction button tests

oliverbarnes committed:
WIP remove needless comments

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Finish removing low-value comments

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Move test

oliverbarnes committed:
Remove low-value tests

oliverbarnes committed:
Merge prediction flow tests

oliverbarnes committed:
Merge refresh model test into model discovery one

oliverbarnes committed:
Use http_client's own FakeHttpClient instead of a custom one

oliverbarnes committed:
Remove schemars feature cfg

oliverbarnes committed:
Remove dumb attempt at ensuring refresh of Ollama models

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Remove remainder test from reverted registry refresh attempt

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

# Conflicts:
#	crates/language_models/src/provider/ollama.rs

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Remove debug logs

oliverbarnes committed:
Don't authenticate ollama on edit prediction registry init

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Simplify test setup response

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Fix cargo machete errors

oliverbarnes committed:
Fix broken test

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Update doc with Ollama Cloud mention

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Add symlink to GPL license

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Fix broken refresh from one of the merges

oliverbarnes committed:
Merge branch 'main' into ollama-inline-completions

oliverbarnes committed:
Fix license symlink

oliverbarnes commented:
@mgsloan ready for another pass, when you get a chance. I _hope_ it's in good shape now 🙏 

There's only my remaining question about FIM model detection in the inline comments. Whether we really want to have multiple http requests to check every single model's template for FIM tokens.

Thanks for the patience.

mgsloan committed:
Merge branch 'main' into ollama-inline-completions

strive1216 commented:
would really love this getting merged

SomeoneToIgnore commented:
I think it's time to close this PR then.
